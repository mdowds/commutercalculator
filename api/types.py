from typing import NamedTuple, Dict, Any, Optional

from api.data import Travelcard
from .data import Station

JourneyResult = NamedTuple('JourneyResult', (
    ('origin', Station),
    ('time', int),
    ('travelcard', Optional[Travelcard]),
))

JourneysToArgs = NamedTuple('JourneysToArgs', (
    ('min_time', int),
    ('max_time', int)
))

JourneyPrice = NamedTuple('JourneyPrice', (
    ('origin', Station),
    ('destination', Station),
    ('travelcard', Optional[Travelcard])
))

OutputDict = Dict[str, Any]
