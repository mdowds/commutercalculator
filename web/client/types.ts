export interface Station {
    readonly id: string,
    readonly name: string,
    readonly position: Position
}

export interface JourneyResult {
    readonly origin: Station,
    readonly journeyTime: number,
    readonly seasonTicket: SeasonTicket
}

export interface SeasonTicket {
    readonly price?: number
}

export interface SelectedFilters {
    minTime?: number,
    maxTime?: number
}

export interface Position {
    readonly lat: number,
    readonly lng: number
}
