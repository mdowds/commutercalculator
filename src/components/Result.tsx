import * as React from "react";
import { JourneyResult, SeasonTickets, Travelcard } from "../types";
import { CSSProperties } from "react";

interface ResultProps {
  readonly result: JourneyResult;
  readonly isSelected: boolean;
  readonly averageHousePrice?: number;
  onSelectResult(selectedResult: JourneyResult): void;
}

export default class Result extends React.Component<ResultProps, {}> {
  travelcardDescription(travelcard: Travelcard) {
    return (
      <div>
        Zone {travelcard.minZone} to {travelcard.maxZone} travelcard: £
        {travelcard.price}
      </div>
    );
  }

  housePriceDescription() {
    const outcode = this.props.result.origin.postcode.split(" ")[0];
    const price = this.props.averageHousePrice
      ? "£" + this.props.averageHousePrice.toLocaleString("en-GB")
      : "Loading";
    return (
      <div>
        Average house price for {outcode}: {price}
      </div>
    );
  }

  makeDetails(result: JourneyResult) {
    const travelcardDiv = result.seasonTickets.travelcard
      ? this.travelcardDescription(result.seasonTickets.travelcard)
      : null;
    const seasonTicketDiv = result.seasonTickets.seasonTicket ? (
      <div>
        Direct season ticket: £{result.seasonTickets.seasonTicket.price}
      </div>
    ) : null;

    const seasonTicketsDescription = this.seasonTicketsHasValues(
      result.seasonTickets
    ) ? (
      <div>
        {travelcardDiv}
        {seasonTicketDiv}
      </div>
    ) : (
      <div>Season ticket not found</div>
    );

    const directionsLink = (
      <span>
        <a href={result.directionsUrl}>View directions on Google Maps</a> (May
        show a different journey time)
      </span>
    );

    return (
      <div>
        {seasonTicketsDescription}
        {this.housePriceDescription()}
        {directionsLink}
      </div>
    );
  }

  seasonTicketsHasValues(seasonTickets: SeasonTickets) {
    return seasonTickets.travelcard || seasonTickets.seasonTicket;
  }

  render() {
    const bgColour = this.props.isSelected ? "#ddd" : "white";

    const divStyle: CSSProperties = {
      padding: 10,
      borderBottom: "1px solid grey",
      backgroundColor: bgColour
    };

    const details = this.props.isSelected ? (
      <div style={{ margin: "10px 0px 0px 30px", fontSize: "75%" }}>
        {this.makeDetails(this.props.result)}
      </div>
    ) : null;

    const icon = this.props.isSelected
      ? "img/chevron-up.svg"
      : "img/chevron-down.svg";

    return (
      <div
        style={divStyle}
        onClick={() => {
          this.props.onSelectResult(this.props.result);
        }}
      >
        <div>
          <div style={{ float: "left", marginRight: 10 }}>
            <img
              alt="Expand results"
              src={icon}
              style={{ width: 20, height: 20 }}
            />
          </div>
          <div style={{ display: "inline-block", width: "65%" }}>
            {this.props.result.origin.name}
          </div>
          <div style={{ float: "right" }}>
            {this.props.result.journeyTime} mins
          </div>
        </div>
        {details}
      </div>
    );
  }
}
