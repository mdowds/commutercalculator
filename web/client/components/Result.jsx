"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
function travelcardDescription(travelcard) {
    return <div>Zone {travelcard.minZone} to {travelcard.maxZone} travelcard: £{travelcard.price}</div>;
}
function makeDetails(result) {
    var travelcardDiv = result.seasonTickets.travelcard ? travelcardDescription(result.seasonTickets.travelcard) : null;
    var seasonTicketDiv = result.seasonTickets.seasonTicket ? <div>Direct season ticket: £{result.seasonTickets.seasonTicket.price}</div> : null;
    var seasonTicketsDescription = <div>
        {travelcardDiv}
        {seasonTicketDiv}
    </div>;
    return seasonTicketsHasValues(result.seasonTickets) ?
        seasonTicketsDescription :
        <div>Season ticket not found</div>;
}
function seasonTicketsHasValues(seasonTickets) {
    return seasonTickets.travelcard || seasonTickets.seasonTicket;
}
function Result(props) {
    var bgColour = props.isSelected ? '#ddd' : 'white';
    var divStyle = {
        padding: 10,
        borderBottom: "1px solid grey",
        backgroundColor: bgColour
    };
    var details = props.isSelected ? <div style={{ margin: '10px 0px 0px 30px', fontSize: '75%' }}>{makeDetails(props.result)}</div> : null;
    var icon = props.isSelected ? "img/chevron-up.svg" : "img/chevron-down.svg";
    return (<div style={divStyle} onClick={function () { props.onSelectResult(props.result); }}>
            <div>
                <div style={{ float: "left", marginRight: 10 }}><img src={icon} style={{ width: 20, height: 20 }}/></div>
                <div style={{ display: "inline-block", width: "65%" }}>{props.result.origin.name}</div>
                <div style={{ float: "right" }}>{props.result.journeyTime} mins</div>
            </div>
            {details}
        </div>);
}
exports.default = Result;
