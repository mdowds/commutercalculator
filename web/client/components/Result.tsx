import * as React from 'react';
import {JourneyResult, Station} from '../types';
import {CSSProperties} from "react";

interface ResultProps {
    readonly result: JourneyResult,
    readonly isSelected: boolean,
    onSelectResult(selectedResult: string): void
}

function makeDetails(result: JourneyResult) {
    return result.seasonTicket ? <div>Season ticket price: Â£{result.seasonTicket.price}</div> : null;
}

export default function Result(props: ResultProps) {

    const bgColour = props.isSelected ? '#ddd' : 'white';

    const divStyle: CSSProperties = {
        padding: 10,
        borderBottom: "1px solid grey",
        backgroundColor: bgColour
    };

    const details = props.isSelected ? <div style={{margin: '10px 0px 0px 30px', fontSize: '75%'}}>{makeDetails(props.result)}</div> : null;

    const icon = props.isSelected ? "img/chevron-up.svg" : "img/chevron-down.svg";

    return (
        <div style={divStyle} onClick={() => {props.onSelectResult(props.result.origin.id)}}>
            <div>
                <div style={{float: "left", marginRight: 10}}><img src={icon} style={{width: 20, height: 20}}/></div>
                <div style={{display: "inline-block", width: "65%"}}>{props.result.origin.name}</div>
                <div style={{float: "right"}}>{props.result.journeyTime} mins</div>
            </div>
            {details}
        </div>
    );
}
