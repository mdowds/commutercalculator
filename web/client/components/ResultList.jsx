"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var Result_1 = require("./Result");
function default_1(props) {
    var entries = props.results.map(function (result) {
        var showDetails = props.selectedResult == result;
        return <Result_1.default key={result.origin.id} result={result} isSelected={showDetails} onSelectResult={props.onSelectResult} averageHousePrice={showDetails ? props.selectedResultHousePrice : undefined}/>;
    });
    var containerStyleDefault = {
        width: "100%",
        background: "white",
        height: "100%",
        overflow: "scroll"
    };
    var containerStyle = props.styles === undefined ? containerStyleDefault : Object.assign(containerStyleDefault, props.styles);
    var loadingIndicator = <div style={{ textAlign: "center", marginTop: 50 }}>Results loading</div>;
    return (<div style={containerStyle}>
            {props.isLoading ? loadingIndicator : entries}
        </div>);
}
exports.default = default_1;
