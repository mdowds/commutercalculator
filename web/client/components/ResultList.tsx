import * as React from 'react';
import {default as Result} from './Result';
import {JourneyResult} from '../types';
import {CSSProperties} from "react";

interface ResultListProps {
    readonly results: Array<JourneyResult>,
    selectedResult?: JourneyResult,
    readonly styles?: CSSProperties,
    readonly isLoading: boolean,
    onSelectResult(selectedResult: JourneyResult): void;
}

export default function(props) {
    const entries = props.results.map((result) => {
        const showDetails = props.selectedResult == result;

        return <Result key={result.origin.id} result={result} isSelected={showDetails} onSelectResult={props.onSelectResult} />
    });

    const containerStyleDefault: CSSProperties = {
        width: "100%",
        background: "white",
        height: "100%",
        overflow: "scroll"
    };

    const containerStyle = props.styles === undefined ? containerStyleDefault : Object.assign(containerStyleDefault, props.styles);

    const loadingIndicator = <div style={{textAlign: "center", marginTop: 50}}>Results loading</div>;

    return (
        <div style={containerStyle}>
            {props.isLoading ? loadingIndicator : entries}
        </div>
    );
}
