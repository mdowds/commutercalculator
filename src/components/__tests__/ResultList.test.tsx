import * as React from 'react';
import * as renderer from 'react-test-renderer';
import {} from 'jest';
import { shallow } from 'enzyme';

import ResultList from '../ResultList';
import Result from '../Result';
import {JourneyResult} from "../../types";

test("ResultList renders the container", () => {
    const list = renderer.create(<ResultList results={[]} isLoading={false} onSelectResult={() => {}}/>);
    expect(list.toJSON()).toMatchSnapshot();
});

test("ResultList renders the container with custom styles", () => {
    const list = renderer.create(<ResultList results={[]} isLoading={false} styles={{height: 200}}  onSelectResult={() => {}} />);
    expect(list.toJSON()).toMatchSnapshot();
});

test("ResultList renders the loading indicator", () => {
    const list = renderer.create(<ResultList results={[]} isLoading={true} onSelectResult={() => {}} />);
    expect(list.toJSON()).toMatchSnapshot();
});

test("ResultList renders results", () => {
    const list = shallow(<ResultList results={[mockResult("foo", 10), mockResult("bar", 20)]} isLoading={false} onSelectResult={() => {}} />);
    expect(list.find(Result).length).toEqual(2);
});

test("ResultList renders details for one result when selectedResult is set", () => {
    const res = mockResult("foo", 10);
     const list = shallow(<ResultList results={[res, mockResult("bar", 20)]} isLoading={false} onSelectResult={() => {}} selectedResult={res} />);
     expect(list.find('[isSelected=true]').length).toEqual(1);
});

function mockResult(id, time): JourneyResult {
    return {
        origin: {id: id, name: id, position: {lat: 0, lng: 0}, postcode: "E1 1A"},
        directionsUrl: "",
        journeyTime: time,
        seasonTickets: {}
    }
}
