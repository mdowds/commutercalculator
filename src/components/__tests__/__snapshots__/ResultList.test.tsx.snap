// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResultList renders the container 1`] = `
<div
  style={
    Object {
      "background": "white",
      "height": "100%",
      "overflowY": "scroll",
      "width": "100%",
    }
  }
/>
`;

exports[`ResultList renders the container with custom styles 1`] = `
<div
  style={
    Object {
      "background": "white",
      "height": 200,
      "overflowY": "scroll",
      "width": "100%",
    }
  }
/>
`;

exports[`ResultList renders the loading indicator 1`] = `
<div
  style={
    Object {
      "background": "white",
      "height": "100%",
      "overflowY": "scroll",
      "width": "100%",
    }
  }
>
  <div
    style={
      Object {
        "marginTop": 50,
        "textAlign": "center",
      }
    }
  >
    Results loading
  </div>
</div>
`;
